src/edu/csus/ecs/pc2/core/LanguageUtilities.java

```diff
@@ -47,39 +62,24 @@ public static Language guessLanguage(IInternalContest myContest, String filename
     public static Language matchFirstLanguage(IInternalContest inContest, String extension) {
         Language[] lang = inContest.getLanguages();
 
-        // Alas guessing 
-        if ("cpp".equals(extension)) {
-            extension = "C++";
-        }
-
-        if ("cc".equals(extension)) {
-            extension = "C++";
-        }
-
-        if ("C".equals(extension)) {
-            extension = "C++";
-        }
-
-        if ("c++".equals(extension)) {
-            extension = "C++";
-        }
-
-        if ("py".equals(extension)) {
-            extension = "Python";
-        }
-
-        if ("cs".equals(extension)) {
-            extension = "Mono";
-        }
+        /**
+         * Partial or complete display name
+         */
+        String displayName = extension;
 
-        if ("pl".equals(extension)) {
-            extension = "Perl";
+        for (String[] row : LANGUAGE_LIST) {
+            String fileExtension = row[0];
+            String dispName = row[1];
+            
+            if (fileExtension.equals(extension)) {
+                displayName = dispName;
+            }
```

Comment:
I might put a 'break' after setting displayName; no point in going through the rest of the loop.  But, it's not an approval breaker.